test_conformance/subgroups/workgroup_common_templates.h

```diff
@@ -0,0 +1,585 @@
+                {
+                    for (i = 0; i < n; ++i)
+                    {
+                        if (Which == 2 && i >= NON_UNIFORM)
+                        {
+                            break; // non uniform case - only first 4 workitems
+                                   // should broadcast. Others are undefined.
+                        }
+                        rr = my[ii + i]; // read device outputs for work_item in
+                                         // the subgroup
+                        if (!compare(rr, tr))
+                        {
+                            if (Which == 0)
+                            {
+                                log_error("ERROR: sub_group_broadcast(%s) "
+                                          "mismatch for local id %d in sub "
+                                          "group %d in group %d\n",
+                                          TypeManager<Ty>::name(), i, j, k);
+                            }
+                            if (Which == 2)
+                            {
+                                log_error("ERROR: "
+                                          "sub_group_non_uniform_broadcast(%s) "
+                                          "mismatch for local id %d in sub "
+                                          "group %d in group %d\n",
+                                          TypeManager<Ty>::name(), i, j, k);
+                            }
+                            return -1;
+                        }
+                    }
+                }
+            }
+            x += nw;
+            y += nw;
+            m += 4 * nw;
+        }
+        return 0;
+    }
+};
+
+template <typename Ty, int Which> struct OPERATION;
+
+template <typename Ty> struct OPERATION<Ty, 0>
+{
+    static Ty calculate(Ty a, Ty b) { return a + b; }
+};
+template <typename Ty> struct OPERATION<Ty, 1>
+{
+    static Ty calculate(Ty a, Ty b) { return a > b ? a : b; }
+};
+template <typename Ty> struct OPERATION<Ty, 2>
+{
+    static Ty calculate(Ty a, Ty b) { return a < b ? a : b; }
+};
+template <typename Ty> struct OPERATION<Ty, 3>
+{
+    static Ty calculate(Ty a, Ty b) { return a * b; }
+};
+template <typename Ty> struct OPERATION<Ty, 4>
+{
+    static Ty calculate(Ty a, Ty b) { return a & b; }
+};
+template <typename Ty> struct OPERATION<Ty, 5>
+{
+    static Ty calculate(Ty a, Ty b) { return a | b; }
+};
+template <typename Ty> struct OPERATION<Ty, 6>
+{
+    static Ty calculate(Ty a, Ty b) { return a ^ b; }
+};
+template <typename Ty> struct OPERATION<Ty, 7>
+{
+    static Ty calculate(Ty a, Ty b) { return a && b; }
+};
+template <typename Ty> struct OPERATION<Ty, 8>
+{
+    static Ty calculate(Ty a, Ty b) { return a || b; }
+};
+template <typename Ty> struct OPERATION<Ty, 9>
+{
+    static Ty calculate(Ty a, Ty b) { return !a ^ !b; }
+};
+
+static float to_float(subgroups::cl_half x) { return cl_half_to_float(x.data); }
+
+static subgroups::cl_half to_half(float x)
```

Comment:
These templates could probably be better suited for operator overloading on subgroups::cl_half. Would allow code sharing with other types that are able to use the regular operators.

Reply:
Can this stay as is ?

Reply:
I won't block the merge myself. We can check if others have differing opinions.